docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+Given a list of role variables and their properties (e.g. if a variable is required, a list of accepted values)
+can change over time, it is important to keep the checks in sync with the `default/main.yml`.
+The following checks can be generated:
+
+```yaml
+- name: "Fail if r_openshift_node_deployment_type is not defined"
+  fail:
+    msg: "r_openshift_node_deployment_type must be specified for this role"
+  when:
+  - r_openshift_node_deployment_type is not defined
+
+- name: Fail if invalid r_openshift_node_deployment_type provided
+  fail:
+    msg: "r_openshift_node_deployment_type can only be set to a single value from ['origin', 'openshift-enterprise']"
+  when:
+  - r_openshift_node_deployment_type is defined
+  - r_openshift_node_deployment_type not in ['origin', 'openshift-enterprise']
+```
+
+All the checks can be stored into `tasks/pre_checks.yml` file and included in the main task file.
+
+#### Propagation of values to role dependencies
+
+A role can depend on another role that has its own set of role variables. There are two ways how to propagate values to role dependencies. First approach requires a caller has a knowledge of all role dependency variables:
+
+```yaml
+roles:
+- role: etcd
+  r_etcd_client_port: 2379
+  r_etcd_peer_port: 2380
+  r_etcd_common_etcd_runtime: docker
+  r_etcd_common_embedded_etcd: false
+```
+
+Here the `etcd` role depends on `etcd_common` role.
+The second approach is to duplicate all role dependency variables:
+
+```yaml
+roles:
+- role: etcd
+  r_etcd_client_port: 2379
+  r_etcd_peer_port: 2380
+  r_etcd_etcd_runtime: docker
+  r_etcd_embedded_etcd: false
+```
+
+and write all the duplicated role variables into their corresponding
+dependency variables in `etcd/meta/main.yml`:
+
+```yaml
+dependencies:
+- role: etcd_common
+  r_etcd_common_etcd_runtime: "{{ r_etcd_etcd_runtime }}"
+  r_etcd_common_embedded_etcd: "{{ r_etcd_embedded_etcd }}"
+```
+
+One can use `include_role` instead of `roles`:
+
+```yaml
+include_role:
+  name: etcd
+vars:
+  r_etcd_client_port: 2379
+  r_etcd_peer_port: 2380
+  r_etcd_etcd_runtime: docker
+  r_etcd_embedded_etcd: false
+```
+
+From the point of view of the caller the number of role variables
+to be set does not change in both approaches.
+
+**Advantages** of the second approach:
+* role dependencies and role variables of dependencies can change transparently
+* caller does not need any knowledge of role variables of dependencies
+
+**Disadvantages**:
+* role variables of dependencies has to be duplicated
+* one must not forget to wire all duplicated variables inside `meta/main.yml` file (this step can be automatically generated)
+* optional undefined variables can not be set this way (TODO(jchaloup): or can they? is that even needed, maybe all roles with optional undefined role variables should not be dependencies of another role)
+
+### Local variables
+
+All variables set inside a role has to be prefixed with `l_` string.
```

Comment:
`l_` is still not clear because it could end up being scoped from anywhere previous in the Ansible execution.  If we want something to be local, the best we could do is use vars/main.yml in the role. See http://docs.ansible.com/ansible/latest/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable